activesupport/lib/active_support/core_ext/time/calculations.rb

```diff
@@ -47,8 +47,10 @@ def at_with_coercion(*args)
       # Time.at can be called with a time or numerical value
       time_or_number = args.first
 
-      if time_or_number.is_a?(ActiveSupport::TimeWithZone) || time_or_number.is_a?(DateTime)
-        at_without_coercion(time_or_number.to_f).getlocal
```

Comment:
The gist of this pull request to get rid of this `to_f` that causes the loss of nanosecond precision.

This could have probably be achieved by using `time_or_number.to_r` for `ActiveSupport::TimeWithZone`, but that would make it the whole trickery around `Time.at` even harder to understand.

Reply:
I think calling `to_r` ourselves is preferable to implementing `to_int` just so that Ruby will call `to_r` internally.

Reply:
That would only solve the problem with precision loss in `Time.at`.
I'm not 100% certain that's the only use case that `Time` uses to distinguish a `String` from a Time-like objects.
[`num_exact`](https://github.com/ruby/ruby/blob/0fb1a67bda8df28ed954f4c4eb7782e11a35beb9/time.c#L525) which performs `respond_to?(:to_int)` check, seems to be used outside of `Time.at`. Can't tell for sure if that could lead to loss of the precision, but why risk and work around instead of following the protocol for Time-like object that Ruby dictates?

Reply:
Several other core methods also check whether their arguments respond to `to_int`: https://github.com/ruby/spec/search?q=%22should_receive+to_int%22

Allowing those methods to accept `ActiveSupport::TimeWithZone` when they couldn't before seems risky. Applications may also be using `respond_to?(:to_int)` to detect numeric values, which would also break with this change.

Calling `to_r` explicitly resolves the known problem (`Time.at` losing precision), with minimal risk of unwanted side-effects. Implementing `to_int` may solve other (hypothetical) problems, but the higher risk isn't worth it, IMO.

Reply:
I agree that `Time` itself doesn't respond to `to_int` (it responds `to_i`), and this makes this Ruby interface a bit awkward. However, what concerns you exactly?
Let's see those `to_int` usages one by one (it would probably be easier to check all `ActiveSupport::TimeWithZone` usages in Rails):
1. `Integer`'s `<<` and `>>` shifts.
If a `Time` object is passed as an argument, it's an `undefined method 'to_int'`, but if it's `AS::TWZ`, the exception is different, `TypeError: can't convert ActiveSupport::TimeWithZone to Integer (ActiveSupport::TimeWithZone#to_int gives NilClass)`. For me, the second is even more descriptive than the first, WDYT?
2. For usages as Array's element reference index argument, it's somewhat similar: `TypeError: no implicit conversion of Time into Integer` vs `TypeError: can't convert ActiveSupport::TimeWithZone to Integer (ActiveSupport::TimeWithZone#to_int gives NilClass)`.
3... The same for `Process.getrlimit`, `String#[]=`, `Array#drop`, `Array#sample`, `String#truncate`, `Process.setrlimit`, `Array#rotate`, `String#setbyte`, `Array#cycle`, `String#sum`, `Array.new`
5. `Array#pack` with Time/AS::TWZ arguments seem to behave similarly to the above given an integer directive pack format, e.g. `'Q'`.
4. Doesn't really apply to `Array#hash`, since it delegates to `hash` and expects `to_int` to be called on that hash, and Time-like is unlikely to have a chance to be returned there.

The list above makes the results from the first two pages of the search you've referenced.
So, from what I can see so far, there is only a difference in the exception message where `AS::TWZ` is used instead of `Time`, and it's not in a way more cryptic with `AS::TWZ` than it is with `Time`.

That makes me think that if a `Time` can't be used as an argument somewhere, the addition of `to_int` to `AS::TWZ` doesn't make a culprit of using `AS::TWZ` there, there's an exception clearly saying that it should not be used as such.

Do you have some real cases (in a way that `AS::TWZ` can't be used in place of `Time`, or where `AS::TWZ` can be used somewhere where the use of `Time` would raise an error)?
I can check for the usages of `AS::TWZ` in search for places where `to_r` should be called and it isn't.
This way, we'll have more information on hand to be able to make a more weighted decision if we want to locally work around by calling `to_r` in Rails and recommend doing so for other projects in Rails ecosystem, or to follow the given Ruby Time-like protocol and implement an empty `to_int` flag-method.

If there's any real problem, and Time-like objects that adhere to Time-like Ruby object protocol involve in a risk of weird behavioral difference between Time and Time-like (`AS::TWZ` in our case), we can open a discussion in Ruby issue tracker, and find a way how to avoid such different behavior.

Does it work for you, @eugeneius ?

Reply:
My concern is that `ActiveSupport::TimeWithZone` isn't just used internally, it's part of Rails' [public API](https://api.rubyonrails.org/classes/ActiveSupport/TimeWithZone.html). Users interact with it pretty frequently, e.g. any time they call `created_at` on a record. I don't want to change its API in a way that's nonsensical for end users when there's an alternative way to get the same result without doing so.

Also, this may be splitting hairs, but I don't believe Ruby actually has a protocol for Time-like objects: rather, `respond_to?(:to_int)` is a protocol for numeric objects, and `Time.at` has some special behaviour for those. Implementing that protocol for `ActiveSupport::TimeWithZone`—which is not numeric—is undesirable, even if it results in the behaviour we want.

Reply:
I understand that.
At the same time, the [documentation](https://api.rubyonrails.org/classes/ActiveSupport/TimeWithZone.html) explicitly states that:

> `TimeWithZone` instances implement the same API as Ruby `Time` instances, so that `Time` and `TimeWithZone` instances are interchangeable.

This boils down to our dilemma since `Time` doesn't implement `to_int`, but in order to be interchangeable, `TimeWithZone` should implement it.

I might be exagerrating when mentioning "protocol", let me know what conclusion you make keeping the following in mind:

1. https://bugs.ruby-lang.org/issues/17131#note-1
> Assuming that the value passed to `Time#at` responds to `to_r` and returns a rational and responds to `to_int`, the precision specified by the rational should be kept.

(this doesn't explicitly states "and **only** if it responds to `to_int`", but I assume this is the case).

https://bugs.ruby-lang.org/projects/ruby-master/repository/git/revisions/7e1fddba4a609cb7bf4a696eccd892e68753bb21
> States Time.at expects rational-like argument to respond to #to_int

2. 
https://github.com/ruby/ruby/blob/520a734ad9c7348f4e4858ee24640f42c88fd389/time.c#L525
```C
            /* test to_int method availability to reject non-Numeric
             * objects such as String, Time, etc which have to_r method. */
```
The reason it lists `Time` along with `String` here is that there's a special case for `Time`, at least in `Time.at` https://github.com/ruby/ruby/blob/520a734ad9c7348f4e4858ee24640f42c88fd389/time.c#L2849
```C
else if (IsTimeval(time)) {
 // COPY the underlying time object data
```


As you can see, the following is not exactly true:
> respond_to?(:to_int) is a protocol for numeric objects

Reply:
My position hasn't changed from:

> Calling `to_r` explicitly resolves the known problem (`Time.at` losing precision), with minimal risk of unwanted side-effects. Implementing `to_int` may solve other (hypothetical) problems, but the higher risk isn't worth it, IMO. 

If there are other problems that would be solved by implementing `to_int`, let's discuss them concretely.

Reply:
I just merged https://github.com/rails/rails/pull/40448, which adds the explicit `to_r` conversion, so that https://github.com/rails/rails/issues/38831 will be fixed in the next release.

If you rebase this to add the DateTime fix on top, I'd be happy to merge it too.